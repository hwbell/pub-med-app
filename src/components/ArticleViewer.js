import React from 'react';
import '../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import { Button, Modal, ModalBody, ModalFooter } from 'reactstrap';

// this form is toggled visible by the 'add to collection' button on the search page
class ArticleViewer extends React.Component {
  constructor(props) {
    super(props);

    this.closeModal = this.closeModal.bind(this);
  }

  renderArticlePreview() {
    let { title, pmcid, pmid, authorString, journalInfo, abstractText, keywordList, fullTextUrlList, pubYear } = this.props.article;


    // authors, date, journal, keywords as subtitles with values below
    let headings = [
      {
        name: 'Authors',
        text: authorString
      },
      {
        name: 'Date Published',
        text: !!journalInfo ? journalInfo.printPublicationDate : pubYear
      },
      {
        name: 'Journal',
        text: !!journalInfo ? journalInfo.journal.title : ''
      },
      {
        name: 'Keywords',
        text: !!keywordList ? keywordList.keyword.join(', ') : ''
      },
      {
        name: 'Abstract',
        text: !!abstractText ? abstractText : ''
      }
    ]

    let ids = [
      {
        name: 'PMID',
        text: pmid
      },
      {
        name: 'PMCID',
        text: pmcid
      }
    ];

    // get urls for the article availablitity if we have them
    let urlList;
    if (fullTextUrlList && fullTextUrlList.fullTextUrl.length > 0) {
      urlList = fullTextUrlList.fullTextUrl.map((articleObj) => {
        // console.log(articleObj)
        let { availabilityCode, url } = articleObj;
        return {
          availabilityCode,
          url
        }
      })
    } else {
      urlList = [];
    }

    return (
      <div className="">

        <Button color="link" className="float-right nav-link" onClick={this.props.toggle}>
          <i className="fas fa-times"></i>
        </Button>

        <p className="preview-title">
          {title}
        </p>

        {/* render the headings / text we have values for */}
        {headings.map((heading, i) => {

          if (!!heading.name && !!heading.text) {
            return (
              <div key={i}>
                <p className="preview-subtitle">{heading.name}</p>
                <p className="preview-text">{heading.text}</p>
              </div>
            )
          } else {
            return null;
          }

        })}

        {/* render the links to the article. They'll link externally to the pdf / access information */}
        {urlList.length > 0 &&
          <div>

            <p className="preview-subtitle">Full Text:</p>

            <div>
              {urlList.map((link, i) => {

                // demand code and link
                if (!!link.availabilityCode && !!link.url) {
                  // set fa class to book or dollar sign
                  let iconClass = link.availabilityCode === 'OA' ? 'fas fa-book-open' : 'fas fa-dollar-sign';

                  return (
                    <a key={i} target="_blank" href={link.url}>
                      <i className={`${iconClass} preview-link`}></i>
                    </a>
                  )

                } else {
                  return null;
                }
              })}
            </div>

          </div>}


        {/* render the identifiers we have values for */}
        <div className="float-right">
          {ids.map((id, i) => {

            if (!!id.name && !!id.text) {
              return (
                <p key={i} className="preview-subtitle">{`${id.name}: ${id.text}`}</p>
              )
            } else {
              return null;
            }
          })}
        </div>
      </div>
    )
  }

  closeModal() {
    this.props.toggle();
  }

  render() {
    // console.log(this.props.article)

    return (
      <Modal contentClassName="outline preview" size="xl" centered={true}
        style={styles.modal}
        isOpen={this.props.isVisible}
        toggle={this.props.toggle}>

        {/* <ModalHeader toggle={this.props.toggle}>Your Collections</ModalHeader> */}
        <ModalBody style={{ margin: '10px', marginBottom: '0px' }}>

          {this.renderArticlePreview()}

        </ModalBody>

      </Modal>

    );
  }
}

const styles = {
  content: {
    height: '100%',
  },
  modal: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  }
}

export default ArticleViewer;
