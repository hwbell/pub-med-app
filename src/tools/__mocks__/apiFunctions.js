const fetch = require('node-fetch')

// query and return based on date / citations
export async function getArticles(query, sortParam) {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  // let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;

  // let headers = {}
  // headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let list = [];
  for (let i = 0; i < 50; i++) {
    list.push({
      title: `Article ${i + 1}`,
      pmid: `PMID10000${i + 1}`
    })
  }

  let articles = await Promise.resolve({
    resultList: {
      result: list
    }
  })

  return articles;
}

// parses the result of a search query to a simple array of the article titles
export function parseSearchToTitlesArray(searchResults) {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })

  return titlesArray;
}