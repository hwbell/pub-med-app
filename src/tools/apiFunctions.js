const fetch = require('node-fetch')

// query and return based on date / citations
const getArticles = async (query, sortParam) => {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;
  
  let headers = {}
  headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let articles = await fetch(searchUrl, {
    method: 'POST',
    headers
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return articles;
}

// parses the result of a search query to a simple array of the article titles
const parseSearchToTitlesArray = (searchResults) => {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })
  return titlesArray;
}

// let results = getArticles('cancer').then((response) => {
//   console.log(response.resultList.result.slice(0,2))
// })

module.exports = {
  getArticles,
  parseSearchToTitlesArray,
}
