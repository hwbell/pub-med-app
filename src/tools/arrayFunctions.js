// this function will remove an article from a specific collection in the collections array
const removeArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles = collection.articles.filter(item => item.id !== article.id)
    }

  });

  return collections;
}

// this function will add an article to a specific collection
const addArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles.push(article);
    }

  });

  return collections;
}

module.exports = {
  removeArticle,
  addArticle
}