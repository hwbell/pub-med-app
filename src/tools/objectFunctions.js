
// this comibines two objects for the keys specified, giving priority to obj1
const combineObjects = (obj1, obj2, keys) => {

  if(!obj1 || !obj2) {
    return console.log('object argument missing')
  }
  if(!keys || !keys.length) {
    return console.log('keys argument missing')
  }

  let combined = {};

  keys.forEach((key) => {
    if (obj1[key]) {
      combined[key] = obj1[key];
    } else if (obj2[key]){
      combined[key] = obj2[key];
    } else {
      combined[key] = '';
    }
  })

  return combined;
}

module.exports = {
  combineObjects
}