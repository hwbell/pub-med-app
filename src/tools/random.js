let list = [];
for (let i = 0; i < 50; i++) {
  list.push({
    title: `Article ${i + 1}`,
    pmid: `PMID10000${i}`
  })
}

console.log(list)