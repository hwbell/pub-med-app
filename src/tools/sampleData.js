
// set some sample data for tests

export function getSampleCollections() {
  let collections = [];
  for (let i = 0; i < 10; i++) {
    collections.push({
      name: `collection ${i + 1}`,
      articles: [
        {
          name: `article ${i + 1}`,
          name: `article ${i + 2}`,
          name: `article ${i + 3}`,
        }
      ]
    })
  }
  return collections;
}