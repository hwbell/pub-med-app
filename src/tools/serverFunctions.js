const collectionServerUrl = process.env.REACT_APP_COLLECTION_SERVER_URL || 'http://localhost:3000/';
const fetch = require('node-fetch');

// this function will actually make a new user OR sign in an existing user, depending on the
// context
export async function signInUser(user, isNewUser) {
  if (!user) {
    return;
  }

  let headers = {}
  headers['content-type'] = 'application/application/json';

  let url = isNewUser ? `${collectionServerUrl}users` : `${collectionServerUrl}users/login`;
  let response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return response;
}

// patches the user's profile
export async function patchUser(patch, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}users/me`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// saves a collection to the server, or patches an exisiting collection on the server
export async function saveCollection(collection, headers, isExisting) {

  // console.log(collection)
  // assign the request params accoring to either PATCH for existing collections or 
  // POST for new collections

  // base url
  let url = `${collectionServerUrl}collections`;
  let method, body;

  if (isExisting) {
    // if its a patch, we only want to send allowed properties.
    let { name, articles } = collection;
    body = {
      name,
      articles
    }
    method = 'PATCH';
    url += `/${collection._id}`;
  } else {
    body = collection;
    method = 'POST';
  }

  let serverResponse = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// deletes user's collection 
export async function deleteCollection(collection, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/${collection._id}`, {
    method: 'DELETE',
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// gets a user's profile / collections / threads
export async function getUserCollections(headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/me`, {
    method: 'GET',
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// patches an existing collection
export async function patchCollection(collection, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/${collection.id}`, {
    method: 'PATCH',
    body: JSON.stringify(collection),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// get the threads on the server, 10 at a time, according to the sortBy param
export async function getPublicThreads(headers, sortBy, page) {

  if (!sortBy || !page) {
    return console.log('Missing sortBy or page parameter.');
  }
  // page represents results 1-10, 11-20, 21-30, etc. 
  let url = `${collectionServerUrl}threads/all/${sortBy}/${page}`;
  let method = 'GET';

  let serverResponse = await fetch(url, {
    method,
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}



// saves a collection to the server, or patches an exisiting collection on the server
export async function saveThread(thread, headers, isComment) {

  // assign the request params accoring to either PATCH for existing thread or 
  // POST for new thread

  // base url
  let url = `${collectionServerUrl}threads`;
  let method, body;

  if (isComment) {
    // if its a patch, we only want to send allowed properties.
    body = {
      user: thread.user,
      text: thread.text,
      add: thread.add
    };
    method = 'PATCH';
    url += `/comments/${thread._id}`;

  } else {
    body = thread;
    method = 'POST';
  }

  let serverResponse = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}








// let user = {
//   "name": makeRandomString(6),
//   "password": makeRandomString(8),
//   "email": `${makeRandomString(6)}@example.com`,
//   "age": 25
// }
// let collection = {
//   "name": "articles about science",
//   "articles": [
//     {
//       "url": "http://nature/bio/moggetarticle1"
//     },
//     {
//       "url": "http://nature/bio/moggetarticle2"
//     },
//     {
//       "url": "http://nature/bio/moggetarticle3"
//     }
//   ]
// }

// let signInAttempt = signInUser(user).then((response) => {

//   console.log(localStorage.token)
// })

// let headers = {
//   Authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZDFhN2VlMTkzYmI0ZTQyZWM1Mjk5MmQiLCJpYXQiOjE1NjIwMTc1MDV9.AT5W-Vu8LssYs3fmH3IpnWSbQFxS0zk8S2WdW-2OAcs',
//   Accept: 'application/json',
//   'Content-Type': 'application/json',
// }
// let saveCollectionAttempt = saveCollection(collection, headers).then((response) => {
//   // 
//   console.log(response);
// })


// module.exports = {
//   signInUser,
//   saveCollection
// }