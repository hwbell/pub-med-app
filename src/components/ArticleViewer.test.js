import React from 'react';
import ArticleViewer from './ArticleViewer';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { getArticles } from '../tools/apiFunctions';

// stubs
// let handleDeleteStub = jest.fn();
// let toggleStub = jest.fn();
// let previewStub = jest.fn();

describe('ArticleViewer', () => {

  let someProps;
  beforeAll(async () => {
    let articles = await getArticles('medicine').then((response) => {
      return response.resultList.result;
    });

    someProps = {
      article: articles[0],
      // handleDelete: handleDeleteStub,
    }
  });

  // tests
  it('renders without crashing', async () => {
    let wrapper = shallow(<ArticleViewer {...someProps}/>);
    // wrapper.update();

  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<ArticleViewer {...someProps}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

})

