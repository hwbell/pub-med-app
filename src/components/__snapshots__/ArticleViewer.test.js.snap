// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleViewer renders correctly 1`] = `null`;
